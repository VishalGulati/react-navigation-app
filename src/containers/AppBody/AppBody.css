.app-body-container {
    width: 90%;
    margin-left: 5%;
    padding: 70px 0;
}