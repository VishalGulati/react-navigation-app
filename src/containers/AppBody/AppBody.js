import React, { Component } from 'react';
import './AppBody.css';
import LeftPanel from '../../components/LeftPanel/LeftPanel';
import RightPanel from '../../components/RightPanel/RightPanel';
import LocationsContext from '../../context/LocationsContext';
import { GOOGLE_API_URL } from '../../config/constants';
import API from '../../axios/AxiosLauncher';
import { URLS } from '../../config/endpoints';


/**
 * The application body, internally split into 2 halves: LeftPanel and RightPanel respectively.
 */
class AppBody extends Component {
    constructor(props) {
        super(props);
        this.state = {
            start: '', drop: '', message: '', messageType: '',
            mapLoaded: false, resetPending: false, showRoute: false,
            route: null, submitBtnLabel: 'Submit'
        };
    }

    handleChange = (key, value) => {
        this.setState({
            [key]: value
        });
    }

    getCords = (obj) => [obj.lat, obj.lng];

    checkForUnsuccessfulMsg = (response) => {
        switch (response.data.status) {
            case 'success':
                return '';
            case 'failure':
                return response.data.error;
            default:
                return response.data.status;
        }
    }

    setMessageInState = (msg, msgType = '') => {
        this.setState({
            message: msg,
            messageType: msgType,
            submitBtnLabel: 'Re-Submit'
        })
    }

    makeRequestForRoute = (endpoint) => {
        API.get(endpoint)
            .then((result) => {
                if (result.data.status === 'in progress') {
                    return this.makeRequestForRoute(endpoint);
                }
                const unsuccessfulMsg = this.checkForUnsuccessfulMsg(result);
                if (unsuccessfulMsg) {
                    this.setMessageInState('Server responded with: ' + unsuccessfulMsg, 'error');
                } else {
                    const { total_distance, total_time } = result.data
                    this.setMessageInState('total distance: ' + total_distance + ' & ' +
                        'total time: ' + total_time);
                    this.setState({ submitBtnLabel: 'Submit', showRoute: true, route: result.data.path });
                }
            })
            .catch((response) => {
                this.setMessageInState('Something went wrong! Please try again in some time.', 'error');
            });
    }

    makeRequestForToken = (origin, destination) => {
        API.post(URLS.submit, { origin, destination })
            .then((result) => {
                const token = result && (result.data && result.data.token);
                const endpoint = URLS.getRoute.replace("{token}", token);
                this.makeRequestForRoute(endpoint);
            })
            .catch((response) => {
                this.setMessageInState('Something went wrong! Please try again in some time.', 'error');
            });
    }

    handleSubmit = (event) => {
        event.preventDefault();
        this.setState({ submitBtnLabel: 'Loading...', message: '', messageType: '' });
        const { start, drop } = this.state;
        if (start && drop) {
            const origin = this.getCords(JSON.parse(JSON.stringify(this.state.start))),
                destination = this.getCords(JSON.parse(JSON.stringify(this.state.drop)));
            this.makeRequestForToken(origin, destination);
        } else {
            this.setMessageInState('Both starting point and drop-off location are mandatory!', 'error');
        }
    }

    resetDone = () => {
        this.setState({ resetPending: false });
    }

    handleReset = (event) => {
        this.setState({ start: '', drop: '', message: '', resetPending: true, submitBtnLabel: 'Submit' });
    }

    initMap = () => {
        this.setState({ mapLoaded: true });
    }

    componentDidMount() {
        window.initMap = this.initMap;
        window.loadJS(`${GOOGLE_API_URL}&libraries=places&callback=initMap`)
    }

    render() {
        return (
            <div className="app-body-container">
                <div className="row">
                    <LocationsContext.Provider value={{
                        mapLoaded: this.state.mapLoaded,
                        updateLocation: this.handleChange,
                        resetPending: this.state.resetPending
                    }}>
                        <LeftPanel
                            handleSubmit={this.handleSubmit}
                            handleReset={this.handleReset}
                            resetDone={this.resetDone}
                            {...this.state} />
                    </LocationsContext.Provider>
                    <RightPanel
                        mapLoaded={this.state.mapLoaded}
                        showRoute={this.state.showRoute}
                        route={this.state.route} />
                </div>
            </div>
        );
    }
}

export default AppBody;