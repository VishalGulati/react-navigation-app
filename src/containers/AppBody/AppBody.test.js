import React from 'react';
import ReactDOM from 'react-dom';
import AppBody from './AppBody';
import { shallow } from 'enzyme';

global.loadJS = jest.fn();
const wrapper = shallow(<AppBody />);

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AppBody />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('Component has container div with class app-body-container', () => {
    expect(wrapper.find('div.app-body-container').length).toEqual(1);
});


it('snapshot testing', () => {
    expect(wrapper).toMatchSnapshot();
});