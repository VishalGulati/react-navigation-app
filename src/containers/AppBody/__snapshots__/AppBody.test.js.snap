// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppBody />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="row"
      >
        <Context.Provider
          value={
            Object {
              "mapLoaded": false,
              "updateLocation": [Function],
            }
          }
        >
          <LeftPanel
            drop=""
            handleReset={[Function]}
            handleSubmit={[Function]}
            mapLoaded={false}
            message=""
            messageType=""
            resetDone={[Function]}
            resetPending={false}
            route={null}
            showRoute={false}
            start=""
            submitBtnLabel="Submit"
          />
        </Context.Provider>
        <RightPanel
          mapLoaded={false}
          route={null}
          showRoute={false}
        />
      </div>,
      "className": "app-body-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Context.Provider
            value={
              Object {
                "mapLoaded": false,
                "updateLocation": [Function],
              }
            }
          >
            <LeftPanel
              drop=""
              handleReset={[Function]}
              handleSubmit={[Function]}
              mapLoaded={false}
              message=""
              messageType=""
              resetDone={[Function]}
              resetPending={false}
              route={null}
              showRoute={false}
              start=""
              submitBtnLabel="Submit"
            />
          </Context.Provider>,
          <RightPanel
            mapLoaded={false}
            route={null}
            showRoute={false}
          />,
        ],
        "className": "row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <LeftPanel
              drop=""
              handleReset={[Function]}
              handleSubmit={[Function]}
              mapLoaded={false}
              message=""
              messageType=""
              resetDone={[Function]}
              resetPending={false}
              route={null}
              showRoute={false}
              start=""
              submitBtnLabel="Submit"
            />,
            "value": Object {
              "mapLoaded": false,
              "updateLocation": [Function],
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "drop": "",
              "handleReset": [Function],
              "handleSubmit": [Function],
              "mapLoaded": false,
              "message": "",
              "messageType": "",
              "resetDone": [Function],
              "resetPending": false,
              "route": null,
              "showRoute": false,
              "start": "",
              "submitBtnLabel": "Submit",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.provider),
            "_context": Object {
              "$$typeof": Symbol(react.context),
              "Consumer": Object {
                "$$typeof": Symbol(react.context),
                "_calculateChangedBits": null,
                "_context": [Circular],
              },
              "Provider": [Circular],
              "_calculateChangedBits": null,
              "_currentRenderer": Object {},
              "_currentRenderer2": null,
              "_currentValue": Object {
                "mapLoaded": false,
                "updateLocation": [Function],
              },
              "_currentValue2": Object {
                "mapLoaded": false,
                "updateLocation": [Function],
              },
              "_threadCount": 0,
            },
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "mapLoaded": false,
            "route": null,
            "showRoute": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
        >
          <Context.Provider
            value={
              Object {
                "mapLoaded": false,
                "updateLocation": [Function],
              }
            }
          >
            <LeftPanel
              drop=""
              handleReset={[Function]}
              handleSubmit={[Function]}
              mapLoaded={false}
              message=""
              messageType=""
              resetDone={[Function]}
              resetPending={false}
              route={null}
              showRoute={false}
              start=""
              submitBtnLabel="Submit"
            />
          </Context.Provider>
          <RightPanel
            mapLoaded={false}
            route={null}
            showRoute={false}
          />
        </div>,
        "className": "app-body-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Context.Provider
              value={
                Object {
                  "mapLoaded": false,
                  "updateLocation": [Function],
                }
              }
            >
              <LeftPanel
                drop=""
                handleReset={[Function]}
                handleSubmit={[Function]}
                mapLoaded={false}
                message=""
                messageType=""
                resetDone={[Function]}
                resetPending={false}
                route={null}
                showRoute={false}
                start=""
                submitBtnLabel="Submit"
              />
            </Context.Provider>,
            <RightPanel
              mapLoaded={false}
              route={null}
              showRoute={false}
            />,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <LeftPanel
                drop=""
                handleReset={[Function]}
                handleSubmit={[Function]}
                mapLoaded={false}
                message=""
                messageType=""
                resetDone={[Function]}
                resetPending={false}
                route={null}
                showRoute={false}
                start=""
                submitBtnLabel="Submit"
              />,
              "value": Object {
                "mapLoaded": false,
                "updateLocation": [Function],
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "drop": "",
                "handleReset": [Function],
                "handleSubmit": [Function],
                "mapLoaded": false,
                "message": "",
                "messageType": "",
                "resetDone": [Function],
                "resetPending": false,
                "route": null,
                "showRoute": false,
                "start": "",
                "submitBtnLabel": "Submit",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": Object {
              "$$typeof": Symbol(react.provider),
              "_context": Object {
                "$$typeof": Symbol(react.context),
                "Consumer": Object {
                  "$$typeof": Symbol(react.context),
                  "_calculateChangedBits": null,
                  "_context": [Circular],
                },
                "Provider": [Circular],
                "_calculateChangedBits": null,
                "_currentRenderer": Object {},
                "_currentRenderer2": null,
                "_currentValue": Object {
                  "mapLoaded": false,
                  "updateLocation": [Function],
                },
                "_currentValue2": Object {
                  "mapLoaded": false,
                  "updateLocation": [Function],
                },
                "_threadCount": 0,
              },
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "mapLoaded": false,
              "route": null,
              "showRoute": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
